src/hooks/useKeyboardShortcut.js

```diff
@@ -0,0 +1,37 @@
+import {useEffect, useRef, useCallback} from 'react';
+import KeyboardShortcut from '../libs/KeyboardShortcut';
+
+/**
+ * Register a keyboard shortcut handler.
+ *
+ * @param {Object} shortcut
+ * @param {Function} callback
+ * @param {Object} [config]
+ */
+export default function useKeyboardShortcut(shortcut, callback, config = {}) {
+    const {captureOnInputs = true, shouldBubble = false, priority = 0, shouldPreventDefault = true, excludedNodes = [], isActive = true} = config;
+
+    const subscription = useRef(null);
+    const subscribe = useCallback(
+        () =>
+            KeyboardShortcut.subscribe(
+                shortcut.shortcutKey,
+                callback,
+                shortcut.descriptionKey,
+                shortcut.modifiers,
+                captureOnInputs,
+                shouldBubble,
+                priority,
+                shouldPreventDefault,
+                excludedNodes,
+            ),
+        [callback, captureOnInputs, excludedNodes, priority, shortcut.descriptionKey, shortcut.modifiers, shortcut.shortcutKey, shouldBubble, shouldPreventDefault],
+    );
+
+    useEffect(() => {
+        const unsubscribe = subscription.current || (() => {});
+        unsubscribe();
+        subscription.current = isActive ? subscribe() : null;
+        return isActive ? subscription.current : () => {};
+    }, [isActive, subscribe]);
+}
```

Comment:
I'm not sure why are we calling manually `unsubscribe();` instead of letting the `unsubscribe` be called during cleanup. I also don't think the reference is needed or the `useCallback`.

I think this could have been simple to implement like this:

```jsx
export default function useKeyboardShortcut(shortcut, callback, config = {}) {
    const {captureOnInputs = true, shouldBubble = false, priority = 0, shouldPreventDefault = true, excludedNodes = [], isActive = true} = config;
    useEffect(() => {
        if (isActive) {
            return KeyboardShortcut.subscribe(
                shortcut.shortcutKey,
                callback,
                shortcut.descriptionKey,
                shortcut.modifiers,
                captureOnInputs,
                shouldBubble,
                priority,
                shouldPreventDefault,
                excludedNodes,
            );
        }
        return undefined;
    }, [isActive, callback, captureOnInputs, excludedNodes, priority, shortcut.descriptionKey, shortcut.modifiers, shortcut.shortcutKey, shouldBubble, shouldPreventDefault]);
}
```
unless I'm missing something?

Also, where we use the hook here:

https://github.com/Expensify/App/blob/63f314c054d82ad0ee5e59ec5349d9b35d656bea/src/components/PopoverMenu/index.js#L52-L62

We didn't use `useCallback` for keeping the same reference of the function. We re-create a new function every render which causes the [useEffect](https://github.com/Expensify/App/blob/63f314c054d82ad0ee5e59ec5349d9b35d656bea/src/hooks/useKeyboardShortcut.js#L31-L36) inside `useKeyboardShortcut` to run every single time after every render.

You can see this problem if you add some console logs:


https://github.com/Expensify/App/assets/87341702/6958385b-5928-4c36-9b35-a4ef41643e04

Reply:
I spent some time trying to fix this by using `useCallback` and `useMemo` to try to keep the same references as much as possible, but I died at [ReportActionCompose](https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionCompose.js). The size and it being a class component doesn't help at all with it :(

Reply:
@aldo-expensify Should we try to export this to expert agency? They might have more luck with this

Reply:
Yes, I think this can be pretty time consuming, it would be great to have their help 🙏 . Thanks!

Reply:
btw, the constant re-subscribe of the shortcut causes this bug: https://github.com/Expensify/App/pull/18883#discussion_r1206010822

Reply:
@aldo-expensify it looks like your implementation would cause a duplicate subscription to be set up every time any of the dependencies to the `useEffect` hook change, without cleaning up the previous subscription. That's why we're manually calling `unsubscribe` in the hook instead of only doing it during the cleanup.

Reply:
Seems like a potential use case for the new `useEvent` React hook.

Reply:
> @aldo-expensify it looks like your implementation would cause a duplicate subscription to be set up every time any of the dependencies to the `useEffect` hook change, without cleaning up the previous subscription. That's why we're manually calling `unsubscribe` in the hook instead of only doing it during the cleanup.

hmm I don't think so, my implementation returns the unsubscribe function (`return KeyboardShortcut.subscribe`). The `useEffect` calls the function during clean up, so it shouldn't be subscribed twice.

Reply:
Oh, I see why I'm wrong here. The cleanup function is called every time the dependencies to the effect change. So if we need to resubscribe, the cleanup (unsubscribe) will be called first, then the effect will run and resubscribe

Reply:
That is my understanding, and I think it works like that... I did some testing adding `console.log`s